rfcs/0000-snark-batching.md

```diff
@@ -0,0 +1,77 @@
+## Summary
+[summary]: #summary
+
+Complicate the SNARK pool so that it can do "batch verification" - verification of several proofs at once.
+
+## Motivation
+[motivation]: #motivation
+
+Verification of new SNARK work becomes a throughput bottleneck at high transaction-per-second configurations (saith [#3981](https://github.com/CodaProtocol/coda/pull/3981)). There is a more efficient "batch verification" that we can use - by verifying N SNARKs at once, we spend `600ms + N*10ms` instead of `N * 600ms` <sup>(citation needed)</sup>. This comes with a catch: if batch verification fails, we only know that _some_ SNARK in there failed to verify, but not which. In the case that a batch fails to verify, we need to do extra work to attribute the failure to a particular sender IP / prover, and know which of the SNARKs we can actually include in the pool. By sending incorrect or fraudulent proofs, an attacker can degrade the throughput of a block producer. Each transaction in the block needs to come with two (or occasionally one) proofs, and if not enough proofs are available, the block producer is forced into leaving profit on the table.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+The proposed mechanism is to have two "batch modes": Batch All and Batch by Prover, and to split provers into two categories: Trusted and Untrusted.
+
+Provers start out Untrusted, and have their proofs checked separately. Proofs from Untrusted provers are batched "by prover": for each prover, all their proofs are verified together in individual batches. If a batch fails to verify, that prover is banned, and none of their proofs will be used. After `TRUST_AMOUNT_PROOFS` proofs, provers are promoted into Trusted. All proofs from trusted provers are verified in a single batch. If that batch fails, the proofs are re-batched "by prover", and the provers of the failing batches are punished.
+
+This changes how SNARK proofs are collected and verified: instead of verifying them immediately as they arrive, we wait for some time before collecting all the received proofs into their respective batches. We'll do this like so:
+
+- Store a queue of pending SNARK work, starting out empty.
+- Upon receipt of the first SNARK work into the queue, start a `MAX_SNARK_DELAY` second timer.
+- When the queue hits `MAX_SNARK_BATCH_LENGTH`, or after the timer expires, clear the queue and start the batch verification process.
+
+The timer ensures that, even when SNARK work is coming in slowly, we aren't adding too much latency into the gossip verification. 
```

Comment:
@emberian is there a need for a `MAX_SNARK_BATCH_LENGTH`? It occurs to me that we could maybe do the following.

Have a state
```ocaml
type verifier_state =
  { mutable out_for_verification : Snark_work.t list option
  ; queue : Snark_work.t Queue.t }
```

- `out_for_verification : Snark_work.t list option` represents the proofs which are currently being processed by the verifier. It is `None` if the verifier is not verifying any proofs at the moment.
- When a new proof comes in put it in the queue. If `out_for_verification = None`, then verify that proof immediately.
- Whenever the verifier returns, flush `queue` into the verifier and move the queue's contents into `out_for_verification`. If the verifier returned successfully, then insert the old `out_for_verification` proofs into the table and regossip them.

Reply:
I guess we could wait `MAX_SNARK_DELAY -  ESTIMATED_VERIFICATION_TIME` in the beginning to get more proofs into that first verification